artemis-commons/src/main/java/org/apache/activemq/artemis/utils/critical/CriticalAnalyzerImpl.java

```diff
@@ -0,0 +1,182 @@
+      components.add(component);
+   }
+
+   @Override
+   public void remove(CriticalComponent component) {
+      components.remove(component);
+   }
+
+   @Override
+   public CriticalAnalyzer setCheckTime(long timeout) {
+      this.checkTime = timeout;
+      return this;
+   }
+
+   @Override
+   public long getCheckTime() {
+      if (checkTime == 0) {
+         checkTime = getTimeout() / 2;
+      }
+      return checkTime;
+   }
+
+   @Override
+   public CriticalAnalyzer setTimeout(long timeout) {
+      this.timeout = timeout;
+      return this;
+   }
+
+   @Override
+   public long getTimeout() {
+      if (timeout == 0) {
+         timeout = TimeUnit.MINUTES.toMillis(2);
+      }
+      return timeout;
+   }
+
+   @Override
+   public CriticalAnalyzer addAction(CriticalAction action) {
+      this.actions.add(action);
+      return this;
+   }
+
+   @Override
+   public void check() {
+      boolean retry = true;
+      while (retry) {
+         try {
+            for (CriticalComponent component : components) {
+
+               int pathReturned = component.isExpired(timeout);
+               if (pathReturned >= 0) {
+                  fireAction(component, pathReturned);
+                  // no need to keep running if there's already a component failed
+                  return;
+               }
+            }
+            retry = false; // got to the end of the list, no need to retry
+         } catch (ConcurrentModificationException dontCare) {
+            // lets retry on the loop
+         }
+      }
+   }
+
+   private void fireAction(CriticalComponent component, int path) {
+      for (CriticalAction action: actions) {
+         try {
+            action.run(component, path);
+         } catch (Throwable e) {
+            logger.warn(e.getMessage(), e);
+         }
+      }
+   }
+
+   @Override
+   public void start() {
+
+      if (!running.tryAcquire()) {
+         // already running
+         return;
+      }
+
+      // we are not using any Thread Pool or any Scheduled Executors from the ArtemisServer
+      // as that would defeat the purpose,
+      // as in any deadlocks the schedulers may be starving for something not responding fast enough
+      thread = new Thread("Artemis Critical Analyzer") {
+         @Override
+         public void run() {
+            try {
+               while (true) {
+                  if (running.tryAcquire(getCheckTime(), TimeUnit.MILLISECONDS)) {
+                     running.release();
+                     // this means that the server has been stopped as we could acquire the semaphore... returning now
+                     break;
+                  }
+                  check();
+               }
+            } catch (InterruptedException interrupted) {
+               // i will just leave on that case
+            }
+         }
+      };
+
+      thread.setDaemon(true);
+
+      thread.start();
+   }
+
+   @Override
+   public void stop() {
```

Comment:
From what I understand start , begins some timer, and stop ends it. If the timer > x action y will occur.

As such if someone in code accidentally forgets to call stop or forgets to put it in a finally block so that when exceptions occur it also stops the timer.